designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,321 @@
+**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 
+### Q: How should Karpenter handle the default or undefined reason case? 
+If a budget reason is unspecified like budgets[2], we will assume this budget applys to all actions that are not specified 
+```yaml
+budgets: 
+  - nodes: 10
+    reasons: [Drift, Consolidation]
+    schedule: "* * * * *"
+  - nodes: 30 
+    schedule: "* * * * *" 
+```
+Meaning that for any actions other than Drift + consolidation, the total amount of disrupting + unhealthy nodes has to be less than 30 for them to trigger disruption.
+
+### Q: How should allowed disruptions be calculated in a multi-reason world? 
+Allowed Disruptions before this RFC are calculated like so 
+AllowedDisruptions = minNodeCountOfAnActiveBudget - unhealthyNodes - totalDisruptingNodes
+With the introduction of Disruption Budget Reasons, this allowedDisruptions equation gets more complex. 
+
+
+There are two equations that make sense when calculating the budgets.
+
+AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes[reason] 
+AllowedDisruptionByReason = minNodeCountOfActiveBudget[reason] - unhealthyNodes - totalDisruptingNodes
+
+The second equation is the one we should opt into.
+
+Take this budget as an example
+```yaml
+budgets: 
+  - nodes: 15 
+    reasons: [Drift, Consolidation]
+    schedule: "* * * * *"
+  - nodes: 10
+    reasons: [Drift]
+    schedule: "* * * * *"
+  - nodes: 30 
+    schedule: "* * * * *" 
+```
+
+First we need to calculate the minNodeCountOfActiveBudget, in this case we note that we set 10 for Drift, since there are two active budgets for drift we take the minimum.
+minNodeCountOfActiveBudget = {
+  Drift: 10 
+  Consolidation: 15 
+  Default: 5
+}
+
+If we use the first equation, we then need to also track in karpenter state the number of nodes currently being disrupted by a given disruption action. 
+disrupting = {
+  Drift: 3 
+  Consolidation: 6
+  Expriation: 3
+  Emptiness: 2
+}
+
+Let say unhealthy nodes are zero. 
+
+**First Equation**
+Drift = 10 - 3 # We can take disruption, because the Drift Bucket isn't full
+Consolidation = 15 - 6 # We can disrupt because 15 - 6 > 0
+Default = 5 - Expriation - Emptiness = 0, # We cannot disrupt any other reasons than consolidation and drift
+**Second Equation** 
+totalDisrupting = 3 + 6 + 5 = 14
+Drift = 10 - 14 # We cannot disrupt due to drift 
+Consolidation = 15 - 14 # We can disrupt due to consolidation at leasrt for one node.
+Default: 5 - 14 # We cannot disrupt any other methods
+
+With the second equation, its much simpler to reason about. A node can be marked for multiple disruption reasons(Emptiness, Drift, Consolidation). SO its challenging to calculate in flight, which exact disruption you will end up hitting. With the second equation, it makes reasoning about how many nodes are getting disrupted simpler for cluster operators.
```

Comment:
@njtran and I talked about this second equation, and came to the conclusion its the simpler behavior to understand. 


When you have 

N Nodepools 
50 Budgets, 

Tracking these behaviors is too complex for users to properly reason about the buckets. 

@garvinp-stripe  what are your thoughts about rather than bucketing by reason, we instead respect the total disrupting count instead? Also makes it easier to address the edge case you and I spoke about

Reply:
At first glance I think that make sense especially since the outcome is more conservative (less nodes disrupted) and untangles the messy logic when nodes are disrupted by multiple reasons. From my perspective this is good. Will let you know if any other thoughts comes up.

Reply:
I like the second equation more as well.

Reply:
ok everyone agreees then we go with second equation.